common/workunit/workflow.cpp

```diff
@@ -613,12 +821,802 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+    unsigned successWfid = item.querySuccess();
+    if(successWfid)
+    {
+        item.setSuccessWfid(insertLogicalPredecessor(successWfid)->queryWfid());
+    }
+    unsigned failureWfid = item.queryFailure();
+    if(failureWfid)
+    {
+        item.setFailureWfid(insertLogicalPredecessor(failureWfid)->queryWfid());
+    }
+}
+void WorkflowMachine::addToItemQueue(unsigned wfid)
+{
+    {
+        CriticalBlock thisBlock(queueCritSec);
+        wfItemQueue.push(wfid);
+    }
+    wfItemQueueSem.signal(1);
+}
+void WorkflowMachine::processDependentSuccessors(CCloneWorkflowItem &item)
+{
+    Owned<IWorkflowDependencyIterator> iter = item.getDependentSuccessors();
+    //item will never be re-executed, so this function is only called once per item
+    if(item.queryWfid() == parentWfid)
+    {
+        //update stop conditions
+        parentReached = true;
+#ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "Reached parent");
+#endif
+        //Evaluate stop conditions. If the workflow has failed, then it needs to check whether there are any pending contingencies
+        checkIfDone();
+    }
+    WorkflowException * e = item.queryException();
+    //MORE: optionally check "alive" - could increase speed, but may introduce race condition
+    for(iter->first();iter->isValid(); iter->next())
+    {
+        unsigned thisWfid = iter->query();
+        CCloneWorkflowItem & cur = queryWorkflowItem(thisWfid);
+        //this must be done even if the workflow item has an exception
+        unsigned numPred = cur.atomicDecNumDependencies();
+        if(e)
+        {
+            bool newBranch = false;
+            {
+                CriticalBlock thisBlock(exceptionCritSec);
+                if(!cur.queryException())
+                {
+                    cur.setException(e);
+                    newBranch = true;
+                }
+            }
+            if(newBranch)
+            {
+                //only process the exception if cur is active
+                if(cur.isActive())
+                {
+                    branchCount++;
+                    addToItemQueue(thisWfid);
+                }
+            }
+        }
+        else
+        {
+            if((numPred == 1) && cur.isActive())
+            {
+                addToItemQueue(thisWfid);
+            }
+        }
+    }
+    if(item.queryException())
+    {
+        //decrement branch count by one, since this item is already on a failed branch
+        branchCount.fetch_add(-1);
+        checkIfDone();
+    }
+}
+void WorkflowMachine::processLogicalSuccessors(CCloneWorkflowItem &item)
+{
+    Owned<IWorkflowDependencyIterator> iter = item.getLogicalSuccessors();
+    for(iter->first();iter->isValid(); iter->next())
+    {
+        unsigned thisWfid = iter->query();
+        CCloneWorkflowItem & cur = queryWorkflowItem(thisWfid);
+        if(!cur.queryContingencyWithin())
+                cur.setContingencyWithin(item.queryContingencyWithin());//item may now be "active"
```

Comment:
The purpose of placing this code early was to prevent an item from being abandoned and then immediately re-added to the task queue. Since this is not a significant time save, I have moved the code to when an item is activated - meaning that only one thread can call it.